### PR TITLE
Publish_L2 / RID LUT Manager

### PR DESCRIPTION
**Pull request by nicHoch**
Preparations for publishing l2 products

also

use of a singelton RidLutManager that pulls in data from the platform api and persiets it into a local csv file

### CODE REVIEW

**Comment by nicHoch** [comment]
end2end is looking good complains about the added header keyword in all fits files

**Review by samaloney — APPROVED** [review]
Does this depend on #382 as that would need to be merged first if so?

Looks ok to me few small queries. What the find.py thing for?

File: stixcore/io/RidLutManager.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        Table
+            the LUT od RIDs and request reasons.
+        """
+        converters = {'_id': np.uint, 'unique_id': np.uint, 'start_utc': datetime,
+                      'duration': np.uint, 'type': str, 'subject': str,
+                      'purpose': str, 'scaling_factor': str, 'ior_id':  str, 'comment': str}
+
+        if update or not file.exists():
+            rid_lut = Table(names=converters.keys(), dtype=converters.values())
+            last_date = date(2019, 1, 1)
```
[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/io/RidLutManager.py:130]: comment added

File: stixcore/io/RidLutManager.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+                    if len(update_lut) < 1:
+                        continue
+                    logger.info(f'found {len(update_lut)} entries')
+                    rid_lut = vstack([rid_lut, update_lut])
+                    # the stix datacenter API is throttled to 2 calls per second
+                    time.sleep(0.5)
+            except Exception as e:
+                logger.error("RID API ERROR")
+                logger.error(e)
```
[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/io/RidLutManager.py:167]: combined

File: stixcore/version_conf.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from pathlib import Path
+
+
+def get_conf_version():
+    try:
+        with open(Path(__file__).parent / "config" / "data" / "common" / "VERSION.TXT") as f:
```
[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/version_conf.py:6]: the version.txt is created by the cron update process of pulling that repo once a day and dumping the git tag into the version file

File: stixcore/io/RidLutManager.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        try:
+            request = self.rid_lut.loc[rid]
+        except KeyError:
+            raise ValueError("can't get scaling factor: no request founds for rid: {rid}")
+        scaling_factor = np.atleast_1d(request['scaling_factor'])
+        if len(scaling_factor) > 1:
+            raise ValueError("can't get scaling factor: to many request founds for rid: {rid}")
+        scf = scaling_factor[0].strip()
+        return 30 if scf == '' else int(float(scf))
```
[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/io/RidLutManager.py:103]: not finding an entry means for the RID no row is avaialable in the table
'' means RID available but no scaling factor set. Other infos request purpose ... might be set

**Comment by nicHoch** [comment]
the find script is a helper CLI tool that scans the given fits folder for files with certain properties. it is not used within the pipeline but will help us to find dedicated fits files for reprocessing/review ...

not many properties implemented so fare but some. Level, version, product_type, OBT_BEG/END ...

File: stixcore/io/fits/processors.py [hunk 5]
```diff
@@ -135,7 +135,8 @@ def generate_common_header(cls, filename, product, *, version=0):
             ('ORIGIN', 'STIX Team, FHNW', 'FHNW'),
             ('CREATOR', 'stixcore', 'FITS creation software'),
             ('VERS_SW', str(stixcore.__version__), 'Version of SW that provided FITS file'),
-            # ('VERS_CAL', '', 'Version of the calibration pack'),
```
[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/io/fits/processors.py:138]: this was already soft removed with the comment before

**Review by samaloney — APPROVED** [review]

File: stixcore/products/level0/scienceL0.py [hunk 6]
```diff
@@ -64,7 +71,7 @@ class NotCombineException(Exception):
     pass
 
 
-class ScienceProduct(GenericProduct, EnergyChannelsMixin):
+class ScienceProduct(GenericProduct, EnergyChannelsMixin, FitsHeaderMixin):
     """Generic science data product class composed of control and data."""
     def __init__(self, *, service_type, service_subtype, ssid, control, data, **kwargs):
         """Create a generic science data product composed of control and data.
```
Comment (nicHoch) [inline stixcore/products/level0/scienceL0.py:77]: @samaloney here the L0 files getting the mixing

File: stixcore/products/level0/scienceL0.py [hunk 7]
```diff
@@ -510,13 +523,17 @@ def from_levelb(cls, levelb, parent=''):
...
                    ssid=packets.ssid,
                    control=control,
                    data=data,
                    idb_versions=idb_versions,
-                   packets=packets)
+                   packets=packets,
+                   history=header_history)
+
+        prod.add_additional_header_keywords(additional_header_keywords)
+        return prod
```
Comment (nicHoch) [inline stixcore/products/level0/scienceL0.py:536]: @samaloney here a list of keywords is added
there is also a  prod.add_additional_header_keyword() if you would like to add a single one
but most likely as the product is created as last step you have to store the keyword in a list during processing

File: stixcore/products/level0/scienceL0.py [hunk 8]
```diff
@@ -594,10 +614,13 @@ def from_levelb(cls, levelb, parent=''):
         triggers = np.array(triggers).reshape(-1, 16).T
 
         if control['compression_scheme_triggers_skm'].tolist() == [[0, 0, 7]]:
-            logger.debug('Unscaling trigger ')
+            factor = RidLutManager.instance.get_scaling_factor(control['request_id'][0])
+            header_history.append(f"trigger descaled with {factor}")
+            logger.debug(f'Unscaling trigger: {factor}')
+            additional_header_keywords.append(("TRIG_SCA", factor, 'used trigger descale factor'))
             triggers, triggers_var = unscale_triggers(
```
Comment (nicHoch) [inline stixcore/products/level0/scienceL0.py:621]: @samaloney here i keep track of one additional keyword

**Pull request merged**